<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Introduction" data-toc-modified-id="Introduction-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Introduction</a></span><ul class="toc-item"><li><span><a href="#Data-preprocessing-steps:" data-toc-modified-id="Data-preprocessing-steps:-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Data preprocessing steps:</a></span><ul class="toc-item"><li><span><a href="#malaria_deaths.csv-->-malaria_processed_deaths.csv:-deaths" data-toc-modified-id="malaria_deaths.csv-->-malaria_processed_deaths.csv:-deaths-1.1.1"><span class="toc-item-num">1.1.1&nbsp;&nbsp;</span><strong>malaria_deaths.csv -&gt; malaria_processed_deaths.csv: <code>deaths</code></strong></a></span></li><li><span><a href="#malaria_deaths_age.csv-->-malaria_deaths_age_processed.csv:-deaths_age" data-toc-modified-id="malaria_deaths_age.csv-->-malaria_deaths_age_processed.csv:-deaths_age-1.1.2"><span class="toc-item-num">1.1.2&nbsp;&nbsp;</span><strong>malaria_deaths_age.csv -&gt; malaria_deaths_age_processed.csv: <code>deaths_age</code></strong></a></span></li><li><span><a href="#malaria_inc.csv-->-malaria_inc_processed.csv:-inc" data-toc-modified-id="malaria_inc.csv-->-malaria_inc_processed.csv:-inc-1.1.3"><span class="toc-item-num">1.1.3&nbsp;&nbsp;</span><strong>malaria_inc.csv -&gt; malaria_inc_processed.csv: <code>inc</code></strong></a></span></li></ul></li></ul></li><li><span><a href="#Are-there-any-differences-in-the-csv-files-processed-using-R-or-python?" data-toc-modified-id="Are-there-any-differences-in-the-csv-files-processed-using-R-or-python?-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Are there any differences in the csv files processed using R or python?</a></span><ul class="toc-item"><li><span><a href="#Import-libraries" data-toc-modified-id="Import-libraries-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Import libraries</a></span></li><li><span><a href="#Import-datasets" data-toc-modified-id="Import-datasets-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Import datasets</a></span></li><li><span><a href="#malaria_deaths_age_processed.csv:-deaths_age" data-toc-modified-id="malaria_deaths_age_processed.csv:-deaths_age-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>malaria_deaths_age_processed.csv: <code>deaths_age</code></a></span><ul class="toc-item"><li><span><a href="#Check-dataset-dimensions" data-toc-modified-id="Check-dataset-dimensions-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Check dataset dimensions</a></span></li><li><span><a href="#dataset-structure-and-null-values" data-toc-modified-id="dataset-structure-and-null-values-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>dataset structure and null values</a></span></li><li><span><a href="#Are-both-dataframes-the-same?" data-toc-modified-id="Are-both-dataframes-the-same?-2.3.3"><span class="toc-item-num">2.3.3&nbsp;&nbsp;</span>Are both dataframes the same?</a></span></li></ul></li><li><span><a href="#malaria_processed_deaths.csv:-deaths" data-toc-modified-id="malaria_processed_deaths.csv:-deaths-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>malaria_processed_deaths.csv: <code>deaths</code></a></span><ul class="toc-item"><li><span><a href="#Check-dataset-dimensions" data-toc-modified-id="Check-dataset-dimensions-2.4.1"><span class="toc-item-num">2.4.1&nbsp;&nbsp;</span>Check dataset dimensions</a></span></li><li><span><a href="#dataset-structure-and-null-values" data-toc-modified-id="dataset-structure-and-null-values-2.4.2"><span class="toc-item-num">2.4.2&nbsp;&nbsp;</span>dataset structure and null values</a></span></li><li><span><a href="#Are-both-dataframes-the-same?" data-toc-modified-id="Are-both-dataframes-the-same?-2.4.3"><span class="toc-item-num">2.4.3&nbsp;&nbsp;</span>Are both dataframes the same?</a></span></li></ul></li><li><span><a href="#malaria_inc_processed.csv:-inc" data-toc-modified-id="malaria_inc_processed.csv:-inc-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>malaria_inc_processed.csv: <code>inc</code></a></span><ul class="toc-item"><li><span><a href="#Check-dataset-dimensions" data-toc-modified-id="Check-dataset-dimensions-2.5.1"><span class="toc-item-num">2.5.1&nbsp;&nbsp;</span>Check dataset dimensions</a></span></li><li><span><a href="#dataset-structure-and-null-values" data-toc-modified-id="dataset-structure-and-null-values-2.5.2"><span class="toc-item-num">2.5.2&nbsp;&nbsp;</span>dataset structure and null values</a></span></li><li><span><a href="#Are-both-dataframes-the-same?" data-toc-modified-id="Are-both-dataframes-the-same?-2.5.3"><span class="toc-item-num">2.5.3&nbsp;&nbsp;</span>Are both dataframes the same?</a></span></li></ul></li></ul></li></ul></div>

# Introduction

In this current repository, three (3) csv files from [this Github repo](https://github.com/rfordatascience/tidytuesday/tree/master/data/2018/2018-11-13) were preprocessed using python and R. 


## Data preprocessing steps:

### **malaria_deaths.csv -> malaria_processed_deaths.csv: `deaths`**

- Tidy up column names:
  - lowercased all column names
  - renamed column name for `'Deaths - Malaria - Sex: Both - Age: Age-standardized (Rate) (per 100,000 people)'` to `'age_std_death_rate'`  
  
- Imputed the missing `'Code'` values for 'Wales', 'England', 'Scotland', 'Northern Ireland' as 'GBR' as they represent the same geographic region of the United Kingdom   

- Added a new column `'entity_type'` to allow subsetting of entity types:  
  - `'Region'` : aggregated geographic regions, 
  - `'SDI`: Socio-demographic Index status  
  - `'Country'`: entities with non-NA `'Code'` values  
  - `'World'`: global aggregate, as indicated by entity value == 'World' 
  
### **malaria_deaths_age.csv -> malaria_deaths_age_processed.csv: `deaths_age`**  

- Dropped the row index column `'Unnamed: 0'` as it does not provide any useful information.  

- Imputed the missing `'Code'` values for 'Wales', 'England', 'Scotland', 'Northern Ireland' as 'GBR' as they represent the same geographic region of the United Kingdom   

- Added a new column `'entity_type'` to allow subsetting of entity types:  
  - `'Region'` : aggregated geographic regions, 
  - `'SDI`: Socio-demographic Index status  
  - `'Country'`: entities with non-NA `'Code'` values  
  - `'World'`: global aggregate, as indicated by entity value == 'World' 

### **malaria_inc.csv -> malaria_inc_processed.csv: `inc`**  

- Tidy up column names:
  - lowercased all column names
  - renamed column name for `'Incidence of malaria (per 1,000 population at risk) (per 1,000 population at risk)'` to `'malaria_incidence_1000'` 
  
- Added a new column `'entity_type'` to allow subsetting of entity types:  
  - `'Region'` : aggregated geographic regions, 
  - `'Country'`: entities with non-NA `'Code'` values,
  - `'World'`: global aggregate, as indicated by entity value == 'World', 
  - `'Income/Demographic`: grouping by income or demographic status. These are entities that contain the following terms: 
    - 'Fragile and conflict affected situations',
    - 'Heavily indebted poor countries (HIPC)',
    - 'Late-demographic dividend',
    - 'Least developed countries: UN classification',
    - 'Low & middle income', 
    - 'Low income', 
    - 'Lower middle income',
    - 'Middle income',
    - 'Pre-demographic dividend',
    - 'Upper middle income'

--- 
# Are there any differences in the csv files processed using R or python?

Let's check if there are any differences in the csv files processed using R or python:

## Import libraries

In [1]:
import glob
import numpy as np
import pandas as pd
from pathlib import Path

## Import datasets

In [2]:
python_processed_folder = Path.cwd().parents[0]/'data/processed'
r_processed_folder = Path.cwd().parents[0]/'data/processed/R'

# get list of absolute file paths for all csv files in the respective folders
python_data_filepaths = [file for file in glob.glob(f"{python_processed_folder}/*.csv")]
r_data_filepaths = [file for file in glob.glob(f"{r_processed_folder}/*.csv")]

print(
    f'Names of csv files in python folder: {[Path(file).stem for file in python_data_filepaths]} \n\
    Names of csv files in R folder: {[Path(file).stem for file in r_data_filepaths]}'
)

print(
    f'File sizes of csv files in python folder (bytes): {[Path(file).stat().st_size for file in python_data_filepaths]}'
)
print(
    f'File sizes of csv files in R folder (bytes): {[Path(file).stat().st_size for file in r_data_filepaths]}'
)

Names of csv files in python folder: ['malaria_deaths_age_processed', 'malaria_deaths_processed', 'malaria_inc_processed'] 
    Names of csv files in R folder: ['malaria_deaths_age_processed', 'malaria_deaths_processed', 'malaria_inc_processed']
File sizes of csv files in python folder (bytes): [1402157, 238252, 19811]
File sizes of csv files in R folder (bytes): [1623009, 268269, 22774]


These are all separate files with different file sizes. 

In [3]:
# csv files processed using python
deaths_age_py = pd.read_csv(python_data_filepaths[0])
deaths_py = pd.read_csv(python_data_filepaths[1])
inc_py = pd.read_csv(python_data_filepaths[2])

# csv files processed using R
deaths_age_r = pd.read_csv(r_data_filepaths[0])
deaths_r = pd.read_csv(r_data_filepaths[1])
inc_r = pd.read_csv(r_data_filepaths[2])

## malaria_deaths_age_processed.csv: `deaths_age`

### Check dataset dimensions

In [4]:
deaths_age_py.shape

(30780, 6)

In [6]:
deaths_age_r.shape

(30780, 6)

### dataset structure and null values

In [7]:
deaths_age_py.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30780 entries, 0 to 30779
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   entity       30780 non-null  object 
 1   code         27000 non-null  object 
 2   year         30780 non-null  int64  
 3   age_group    30780 non-null  object 
 4   deaths       30780 non-null  float64
 5   entity_type  30780 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 1.4+ MB


In [8]:
deaths_age_r.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30780 entries, 0 to 30779
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   entity       30780 non-null  object 
 1   code         27000 non-null  object 
 2   year         30780 non-null  int64  
 3   age_group    30780 non-null  object 
 4   deaths       30780 non-null  float64
 5   entity_type  30780 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 1.4+ MB


### Are both dataframes the same?

In [9]:
deaths_age_py.equals(deaths_age_r)

True

## malaria_processed_deaths.csv: `deaths`

### Check dataset dimensions

In [10]:
deaths_py.shape

(6156, 5)

In [11]:
deaths_r.shape

(6156, 5)

### dataset structure and null values

In [12]:
deaths_py.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6156 entries, 0 to 6155
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   entity              6156 non-null   object 
 1   code                5400 non-null   object 
 2   year                6156 non-null   int64  
 3   age_std_death_rate  6156 non-null   float64
 4   entity_type         6156 non-null   object 
dtypes: float64(1), int64(1), object(3)
memory usage: 240.6+ KB


In [13]:
deaths_r.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6156 entries, 0 to 6155
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   entity              6156 non-null   object 
 1   code                5400 non-null   object 
 2   year                6156 non-null   int64  
 3   age_std_death_rate  6156 non-null   float64
 4   entity_type         6156 non-null   object 
dtypes: float64(1), int64(1), object(3)
memory usage: 240.6+ KB


### Are both dataframes the same?

In [14]:
deaths_py.equals(deaths_r)

False

False means that at least one value is not the same. Checking each column individually:

In [46]:
# entity column
deaths_py.entity.equals(deaths_r.entity)

True

In [47]:
# code column
deaths_py.code.equals(deaths_r.code)

True

In [48]:
# year column
deaths_py.year.equals(deaths_r.year)

True

In [49]:
# age_std_death_rate column
deaths_py.age_std_death_rate.equals(deaths_r.age_std_death_rate)

False

In [50]:
# entity_type column
deaths_py.entity_type.equals(deaths_r.entity_type)

True

Checking the age_std_death_rate column summary statistics:

In [53]:
# python file
deaths_py.age_std_death_rate.describe()

count    6156.000000
mean       16.375074
std        38.606219
min         0.000000
25%         0.000000
50%         0.020219
75%         3.891035
max       229.228881
Name: age_std_death_rate, dtype: float64

In [54]:
# R file
deaths_r.age_std_death_rate.describe()

count    6156.000000
mean       16.375074
std        38.606219
min         0.000000
25%         0.000000
50%         0.020219
75%         3.891035
max       229.228881
Name: age_std_death_rate, dtype: float64

In [55]:
deaths_py.age_std_death_rate.describe() == deaths_r.age_std_death_rate.describe()

count     True
mean      True
std      False
min       True
25%       True
50%       True
75%       True
max       True
Name: age_std_death_rate, dtype: bool

The standard deviation of both files are 38.606219. The differences are likely due to floating point differences in the age_std_death_rate column when parsing the files

In [57]:
# checking the subsetted data
deaths_py.loc[
    (deaths_py.age_std_death_rate != deaths_r.age_std_death_rate),
    'age_std_death_rate'
]

7        7.592172
10       7.939949
16       8.459952
17       8.266378
19       7.817882
          ...    
6107    67.887307
6124    25.170050
6130     4.829650
6131     4.881530
6134     5.001814
Name: age_std_death_rate, Length: 468, dtype: float64

In [58]:
deaths_r.loc[
    (deaths_r.age_std_death_rate != deaths_py.age_std_death_rate),
    'age_std_death_rate'
]

7        7.592172
10       7.939949
16       8.459952
17       8.266378
19       7.817882
          ...    
6107    67.887307
6124    25.170050
6130     4.829650
6131     4.881530
6134     5.001814
Name: age_std_death_rate, Length: 468, dtype: float64

In [61]:
deaths_py.iloc[10,:]  # manually inspecting row index 10

entity                Afghanistan
code                          AFG
year                         2000
age_std_death_rate       7.939949
entity_type               Country
Name: 10, dtype: object

In [62]:
deaths_r.iloc[10,:]  # manually inspecting row index 10

entity                Afghanistan
code                          AFG
year                         2000
age_std_death_rate       7.939949
entity_type               Country
Name: 10, dtype: object

In [63]:
deaths_py.iloc[6107,:]  # manually inspecting row index 6107

entity                   Zambia
code                        ZMB
year                       1995
age_std_death_rate    67.887307
entity_type             Country
Name: 6107, dtype: object

In [64]:
deaths_r.iloc[6107,:]  # manually inspecting row index 6107

entity                   Zambia
code                        ZMB
year                       1995
age_std_death_rate    67.887307
entity_type             Country
Name: 6107, dtype: object

The values are the same; the equals test failed likely due to floating point differences in the age_std_death_rate column when parsing the files

## malaria_inc_processed.csv: `inc`

### Check dataset dimensions

In [36]:
inc_py.shape

(508, 5)

In [37]:
inc_r.shape

(508, 5)

### dataset structure and null values

In [38]:
inc_py.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 508 entries, 0 to 507
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   entity                  508 non-null    object 
 1   code                    400 non-null    object 
 2   year                    508 non-null    int64  
 3   malaria_incidence_1000  508 non-null    float64
 4   entity_type             508 non-null    object 
dtypes: float64(1), int64(1), object(3)
memory usage: 20.0+ KB


In [39]:
inc_r.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 508 entries, 0 to 507
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   entity                  508 non-null    object 
 1   code                    400 non-null    object 
 2   year                    508 non-null    int64  
 3   malaria_incidence_1000  508 non-null    float64
 4   entity_type             508 non-null    object 
dtypes: float64(1), int64(1), object(3)
memory usage: 20.0+ KB


### Are both dataframes the same?

In [40]:
inc_py.equals(inc_r)

True